concent_api/core/tests/test_integration_report_computed_task.py

```diff
@@ -429,3 +429,950 @@ def test_provider_forces_computed_task_report_and_requestor_does_not_respond(sel
         self.assertGreaterEqual(message_from_concent_to_requestor.timestamp, int(dateutil.parser.parse("2017-12-01 11:00:05").timestamp()))
         self.assertLessEqual(   message_from_concent_to_requestor.timestamp, int(dateutil.parser.parse("2017-12-01 11:00:15").timestamp()))
         self.assertEqual(message_ack_from_concent_to_requestor, message_from_concent_to_provider)
+
+    def test_provider_forces_computed_task_report_twice_and_concent_returns_400_error(self):
+        """
+        Tests if on provider ForceReportComputedTask message Concent will return HTTP 400 error
+        if ForceReportComputedTask was already sent by this provider.
+
+        Expected message exchange:
+        Provider -> Concent:    ForceReportComputedTask
+        Provider -> Concent:    ForceReportComputedTask
+        Concent   -> Provider:  HTTP 400 error.
+        """
+
+        # STEP 1: Provider forces computed task report via Concent
+
+        task_to_compute = message.MessageTaskToCompute(
+            timestamp   = int(dateutil.parser.parse("2017-12-01 10:00:00").timestamp()),
+            task_id     = 1,
+            deadline    = int(dateutil.parser.parse("2017-12-01 11:00:00").timestamp()),
+        )
+        serialized_task_to_compute = dump(task_to_compute, PROVIDER_PRIVATE_KEY, REQUESTOR_PUBLIC_KEY)
+
+        force_report_computed_task = message.MessageForceReportComputedTask(
+            timestamp               = int(dateutil.parser.parse("2017-12-01 10:59:00").timestamp()),
+            message_task_to_compute = serialized_task_to_compute,
+        )
+        serialized_force_report_computed_task = dump(force_report_computed_task, PROVIDER_PRIVATE_KEY, CONCENT_PUBLIC_KEY)
+
+        with freeze_time("2017-12-01 10:59:00"):
+            response_1 = self.client.post(
+                reverse('core:send'),
+                data                           = serialized_force_report_computed_task,
+                content_type                   = 'application/octet-stream',
+                HTTP_CONCENT_CLIENT_PUBLIC_KEY = b64encode(PROVIDER_PUBLIC_KEY).decode('ascii'),
+            )
+
+        self.assertEqual(response_1.status_code,  202)
+        self.assertEqual(len(response_1.content), 0)
+
+        # STEP 2: Provider forces computed task report via Concent again
+        with freeze_time("2017-12-01 10:59:00"):
+            response_2 = self.client.post(
+                reverse('core:send'),
+                data                           = serialized_force_report_computed_task,
+                content_type                   = 'application/octet-stream',
+                HTTP_CONCENT_CLIENT_PUBLIC_KEY = b64encode(PROVIDER_PUBLIC_KEY).decode('ascii'),
+            )
+
+        self.assertEqual(response_2.status_code, 400)
+        self.assertEqual(response_2.json(), {"error": "MessageForceReportComputedTask is already being processed for this task."})
```

Comment:
@rwrzesien Please don't test the exact text of error messages. This makes for very brittle tests. It's enough to check i the `error` field is present.

Reply:
@cameel I think it makes a decent test. If someone is changing the implementation, he should also change the test. Also, checking only for `error` field present may be not enough as we are returning different errors for different cases. Also, all strings, including error messages, should be defined as constants and used in both implementation and tests without concern about text change.

Reply:
From my experience, this produces a lot of false positives. I can't remember any case where a test failing due to a tweaked message actually found any breakage. But I remember a lot of cases where I had to tediously go over the tests and find and fix each and every changed piece of text. It does not add any value but forces you to write the text twice.

Also it's hard to test this reliably. Messages often go through translation and you have to expend extra effort to make sure you're comparing with the right version. And I'm not sure if extracting all the strings into a translation table is the best way to combat this. What I've seen more often is something like: `_("This is an error message.")`, where `_()` is a `gettext` call.

And it's the same for the actual client - no one should compare these messages on the client side. They're only a hint for the user. What we really should have are machine-readable error codes.

Reply:
I agree that error codes are better, but currently we don't have those. Also there are custom exception classes which are also cool. And as you described a situation when you go over the tests and find and fix each and every changed piece of text, that is a perfect reason to use constants.

And you can use translation on both constant, and in tests, as they are run with django context with a specific settings which determine language. It is ok to do:
```
ERROR_1 = _("This is an error message.")
```
If you want I can extract strings into constants from whole application, which I think should free us from maintaining them on both sides in the future, but maybe lets do this as separate task.

Reply:
Generally I think we should remove these checks for now. We can reintroduce them when they can be done right. Leaving them as is the worst of both approaches.

I've been through various approaches to this problem (gettext, costants, global message tables, etc.) in various languages and all of them have significant drawbacks. It's even more evident in applications where you have UI and other people (but not necessarily translators) want to be able to edit those strings. My opinion is simply that text is too volatile and it's not a good idea to test it. It's not worth the hassle.

If you want to refactor the code to add constants for all messages, that would be a bit less bad so I'm fine with it, but please let's not leave the text hard-coded like this.

Reply:
@cameel Ok, as we discussed on daily they will be removed by @kbeker .